### ISSUE TITLE
[HOLD for payment 2024-05-09] [HOLD for payment 2024-05-08] HIGH: [API Reliability] [$250] Multiple calls to `OpenReport` on signin

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#39673]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.60-7
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1712180869292309

## Action Performed:
1. Sign in to NewDot

## Expected Result:
Only one `OpenReport` is called since I open it only once

## Actual Result:
It is called twice

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![Screen Shot 2024-04-04 at 8 51 35 PM](https://github.com/Expensify/App/assets/38435837/bb7cba44-7b2f-47ec-b9c1-2bf8e970171f)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0138cc8ab53b98e8c0</li>
        <li>Upwork Job ID: 1777757590524878848</li>
        <li>Last Price Increase: 2024-04-16</li>
<li>Automatic offers: </li>
<ul>
<li>abdulrahuman5196 | Reviewer | 0</li>
<li>bernhardoj | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There are 2 duplicate OpenReport request when opening a report for the first time, for example after sign in.

### What is the root cause of that problem?
The first OpenReport is called when ReportScreen is mounted (`fetchReportIfNeeded`).
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/home/ReportScreen.tsx#L434-L438

The second OpenReport is called when ReportActionsView is also mounted (`openReportIfNecessary`).
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/home/report/ReportActionsView.tsx#L238-L245

Both `openReportIfNecessary` and `openReportIfNecessary` won't call OpenReport if `shouldFetchReport` is false.
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/home/report/ReportActionsView.tsx#L114-L116

If `shouldFetchReport` is true, then the difference between `openReportIfNecessary` and `openReportIfNecessary` is, `fetchReportIfNeeded` won't call OpenReport if the report already exists
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/home/ReportScreen.tsx#L399-L404

but `openReportIfNecessary` will always call it.
https://github.com/Expensify/App/blob/f1502291a14d4e15e04249b241109f146bdb512f/src/pages/home/report/ReportActionsView.tsx#L113-L119

### What changes do you think we should make in order to solve the problem?

First, don't call `openReportIfNecessary` when ReportActionsView is mounted. Then, we need to decide whether we should always call OpenReport or not when ReportScreen is mounted even when we already have the report object. I would say yes so we always get the latest report data every time we visit a chat (and also because we did this in ReportActionsView, so the behavior is unchanged).

**Comment by muttmuure** [comment]
Adding external

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by abdulrahuman5196** [comment]
@bernhardoj I understand both ReportScreen.tsx and ReportActionsView.tsx component mount has OpenReport has in it. 
But the thing i don't understand is, why is this happening only at signin?, since it doesn't happen when we open a chat after that.
I noticied, that OpenReport is called multiple times only when having a skeleton loading. What is the co-relation here?

I support removing the openReport call inside `ReportActionsView`, but wanted to be sure we won't cause any issue since its a super old code.

**Comment by bernhardoj** [comment]
> But the thing i don't understand is, why is this happening only at signin?, since it doesn't happen when we open a chat after that.
> I noticied, that OpenReport is called multiple times only when having a skeleton loading. What is the co-relation here?

<img width="894" alt="image" src="https://github.com/Expensify/App/assets/50919443/28bb9094-6dbc-490d-9a85-33203e933c7b">

I mentioned here that ReportScreen won't call OpenReport if the report object already exists. There is also a `!reportMetadata?.isLoadingInitialReportActions` condition which will be `false` when you open an existing report, but never load it yet. If you never load it yet, then the `isLoadingInitialReportActions` value will be true (default)
https://github.com/Expensify/App/blob/568921fbca73dd27bdb2b4135d75bbede60b3d18/src/pages/home/ReportScreen.tsx#L712-L715

**Comment by abdulrahuman5196** [comment]
@bernhardoj If I disble ReportActionsView OpenReport API call, there isn't any OpenReport API call not made when opening a report.


https://github.com/Expensify/App/assets/46707890/6437cc22-fd4a-43b3-b4c7-5c1d0c29759e

**Comment by bernhardoj** [comment]
@abdulrahuman5196 yep, that's why I propose to remove the existing report condition from the ReportScreen

> Then, we need to decide whether we should always call OpenReport or not when ReportScreen is mounted even when we already have the report object. I would say yes so we always get the latest report data every time we visit a chat (and also because we did this in ReportActionsView, so the behavior is unchanged).

I have updated my proposal to point out which code to remove

**Comment by quinthar** [comment]
What are the next steps?  What's the ETA?

**Comment by muttmuure** [comment]
@abdulrahuman5196 could you take a look at the updated proposal?

**Comment by abdulrahuman5196** [comment]
Hi, The updated proposal seems reasonable. I will some quick testing in my morning and approve if no issues.

**Comment by abdulrahuman5196** [comment]
@bernhardoj 's proposal here https://github.com/Expensify/App/issues/39673#issuecomment-2041127917 looks good on high level and address the issue. IMO we should call OpenReport in only one place to have consistency. 
@bernhardoj I still saw some reports not making OpenReport API call with the mentioned changes but I think that could be fixed in the PR. We should test with different kind of reports as part of PR and make sure nothing breaks.

🎀 👀 🎀
C+ Reviewed

**Comment by flodnv** [comment]
The proposal sounds good to me but 

>@bernhardoj I still saw some reports not making OpenReport API call with the mentioned changes but I think that could be fixed in the PR. We should test with different kind of reports as part of PR and make sure nothing breaks.

Sounds weird and scary, so let's make sure to test this extensively please.

**Comment by bernhardoj** [comment]
PR is ready

cc: @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
@bernhardoj The PR has been reverted due to an issue. Kindly raise a another PR fixing the regression as well.

**Comment by bernhardoj** [comment]
@abdulrahuman5196 new PR is ready

**Comment by muttmuure** [comment]
Awaiting payment date

**Comment by flodnv** [comment]
@muttmuure can we pay and close this?

**Comment by muttmuure** [comment]
Automation has failed, I will create a payment in Upwork tomorrow

**Comment by muttmuure** [comment]
Paid
